Mathlib/Data/Matroid/Basic.lean

```diff
@@ -0,0 +1,884 @@
+  `Base` of `M`. -/
+@[ext] structure Matroid (α : Type _) where
+  /-- `M` has a ground set. -/
+  (ground : Set α)
+  /-- `M` has a predicate `Base` definining its bases -/
+  (Base : Set α → Prop)
+  /-- There is at least one `Base` -/
+  (exists_base' : ∃ B, Base B)
+  /-- For any bases `B,B'` and `e ∈ B \ B'`, there is some `f ∈ B' \ B` for which `B-e+f`
+    is a base -/
+  (base_exchange' : Matroid.ExchangeProperty Base)
+  /-- Every subset `I` of a set `X` for which `I` is contained in a base is contained in a maximal
+    subset of `X` that is contained in a base. -/
+  (maximality' : ∀ X, X ⊆ ground → Matroid.ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+  /-- every base is contained in the ground set -/
+  (subset_ground' : ∀ B, Base B → B ⊆ ground)
+
+namespace Matroid
+
+variable {α : Type _} {M : Matroid α}
+
+attribute [pp_dot] Base
+
+/-- We write `M.E` for the ground set of a matroid `M`-/
+@[pp_dot] def E (M : Matroid α) : Set α := M.ground
+
+@[simp] theorem ground_eq_E (M : Matroid α) : M.ground = M.E := rfl
+
+/-- Typeclass for a matroid having finite ground set. Just a wrapper for `M.E.Finite`-/
+class Finite (M : Matroid α) : Prop where
+  /-- The ground set is finite -/
+  (ground_finite : M.E.Finite)
+
+/-- Typeclass for a matroid having nonempty ground set. Just a wrapper for `M.E.Nonempty`-/
+class Nonempty (M : Matroid α) : Prop where
+  /-- The ground set is nonempty -/
+  (ground_nonempty : M.E.Nonempty)
+
+theorem ground_nonempty (M : Matroid α) [Nonempty M] : M.E.Nonempty :=
+  Nonempty.ground_nonempty
+
+theorem ground_finite (M : Matroid α) [Finite M] : M.E.Finite :=
+  ‹M.Finite›.ground_finite
+
+theorem set_finite (M : Matroid α) [Finite M] (X : Set α) (hX : X ⊆ M.E := by aesop) : X.Finite :=
+  M.ground_finite.subset hX
+
+instance finite_of_finite [@_root_.Finite α] {M : Matroid α} : Finite M :=
+  ⟨Set.toFinite _⟩
+
+/-- A `FiniteRk` matroid is one whose bases are finite -/
+class FiniteRk (M : Matroid α) : Prop where
+  /-- There is a finite base -/
+  exists_finite_base : ∃ B, M.Base B ∧ B.Finite
+
+instance finiteRk_of_finite (M : Matroid α) [Finite M] : FiniteRk M :=
+  ⟨ M.exists_base'.imp (fun B hB ↦ ⟨hB, M.set_finite B (M.subset_ground' _ hB)⟩) ⟩
+
+/-- An `InfiniteRk` matroid is one whose bases are infinite. -/
+class InfiniteRk (M : Matroid α) : Prop where
+  /-- There is an infinite base -/
+  exists_infinite_base : ∃ B, M.Base B ∧ B.Infinite
+
+/-- A `RkPos` matroid is one whose bases are nonempty. -/
+class RkPos (M : Matroid α) : Prop where
+  /-- The empty set isn't a base -/
+  empty_not_base : ¬M.Base ∅
+
+theorem rkPos_iff_empty_not_base : M.RkPos ↔ ¬M.Base ∅ :=
+  ⟨fun ⟨h⟩ ↦ h, fun h ↦ ⟨h⟩⟩
+
+section exchange
+
+variable {Base : Set α → Prop} (exch : ExchangeProperty Base)
+
+/-- A family of sets with the exchange property is an antichain. -/
+theorem antichain_of_exch (hB : Base B) (hB' : Base B') (h : B ⊆ B') : B = B' :=
```

Comment:
Does it make sense to put this in the `ExchangeProperty` namespace? Something like
```suggestion
theorem ExchangeProperty.antichain (hB : Base B) (hB' : Base B') (h : B ⊆ B') : B = B' :=
```
Then you can use this with dot-notation as `exch.antichain hB hB' h`.